### PR TITLE
Implemented transposed (de-) convolutions into Keras

### PR DESCRIPTION
**Pull request by yaringal**

### CODE REVIEW

File: examples/variational_autoencoder_deconv.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+h_decoded = decoder_h(z)
+f_decoded = decoder_f(h_decoded)
+c_decoded = decoder_c(f_decoded)
+x_decoded_mean = decoder_mean(c_decoded)
+
+
+def vae_loss(x, x_decoded_mean):
+    # NOTE: binary_crossentropy expects a batch_size by dim for x and x_decoded_mean, so we MUST flatten these!!!!!
+    x = Flatten()(x)
```
Comment (fchollet) [inline examples/variational_autoencoder_deconv.py]: Use `K.flatten(x)`, layers don't belong in a loss function.

File: examples/variational_autoencoder_deconv.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    x_decoded_mean = Flatten()(x_decoded_mean)
+    xent_loss = objectives.binary_crossentropy(x, x_decoded_mean)
+    kl_loss = - 0.5 * K.mean(1 + z_log_std - K.square(z_mean) - K.exp(z_log_std), axis=-1)
+    return xent_loss + kl_loss
+
+vae = Model(x, x_decoded_mean)
+opt = RMSprop()
```
Comment (fchollet) [inline examples/variational_autoencoder_deconv.py]: If you're using the default RMSprop parameters, you might as well pass it as a string to `compile()`.

File: keras/layers/convolutional.py [hunk 3]
```diff
@@ -379,6 +379,95 @@ def get_config(self):
         return dict(list(base_config.items()) + list(config.items()))
 
 
+class TransposedConvolution2D(Convolution2D):
```
Comment (fchollet) [inline keras/layers/convolutional.py]: I'd rather use `Deconvolution2D` (and an alias `Deconv2D`), to better match the TensorFlow API.

File: keras/utils/np_utils.py [hunk 4]
```diff
@@ -120,3 +120,13 @@ def conv_output_length(input_length, filter_size, border_mode, stride, dilation=
     elif border_mode == 'valid':
         output_length = input_length - dilated_filter_size + 1
     return (output_length + stride - 1) // stride
+
+def conv_input_length(output_length, filter_size, border_mode, stride):
```
Comment (fchollet) [inline keras/utils/np_utils.py]: `conv_output_length` operation is not always reversible, is it?

File: keras/layers/convolutional.py (hunk 3 above)
Comment (EderSantana) [inline keras/layers/convolutional.py]: keras.backed also says deconv2d. Also, deconv is the most popular term, although not the most precise

File: keras/layers/convolutional.py [hunk 5]
```diff
@@ -379,6 +379,95 @@ def get_config(self):
...
+            rows = input_shape[2]
+            cols = input_shape[3]
+        elif self.dim_ordering == 'tf':
+            rows = input_shape[1]
+            cols = input_shape[2]
+        else:
+            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
+
+        rows = conv_input_length(rows, self.nb_row,
+                                  self.border_mode, self.subsample[0])
```
Comment (EderSantana) [inline keras/layers/convolutional.py]: in the deconv layer, the strides actually upscale the image. I believe it will be confusing to keep "subsample".

**Comment by EderSantana** [comment]
Thanks for the PR @yaringal I only missed a unittest.

Also, the `output_shape` needs to be passed by hand, do you think we can infer that but just forcing the user to use `batch_input_shape` when a deconv layer is part of the model? In TF the shape can be recovered with input `get_shape` plus layer `get_output_shape_for`. I noticed some comments where you tried to catch `output_shape` while building the layer, that didn't work?

File: keras/layers/convolutional.py (hunk 5 above)
Comment (fchollet) [inline keras/layers/convolutional.py]: Maybe "strides". I had been considering deprecating `subsample` in favor of `strides` in Conv2d, too.

**Comment by EderSantana** [comment]
guys, there are lots of conflicts in this branch, I'm starting another one from scratch and making sure the vae deconv example works in tensorflow. 

I'll leave the subsample -> stride modification out of this. Also, I'll keep the hand defined `output_shape` while we think of an automatic inference way to do this. 

I'll make the PR as soon as the vae conv example finishes running (its much slower on TF, is it because of the transpose operations?)

**Comment by fchollet** [comment]
To benchmark against TF, use the TF dim ordering.

**Comment by mmmikael** [comment]
> keras.backed also says deconv2d. Also, deconv is the most popular term, although not the most precise

This naming imbroglio is interesting. While this operation has been called "deconvolution"  in several papers and libraries, this is simply a convolution with fractional strides (i.e. interpolation is needed when applying the kernel). The "classical" meaning of deconvolution is to reverse the effect of a convolution which is not necessarily the case here. For convolutions with stride 1, the corresponding deconvolution would not need this deconv... This makes this naming quite miss-leading...

**Comment by yaringal** [comment]
Apologies for not answering - I've been travelling the past few days with no internet access. How should we proceed? adapt code on this PR or on #3273?

**Comment by EderSantana** [comment]
@yaringal you can just get the code from the other PR and paste it here. This way you get proper credit for you contributions

**Comment by yaringal** [comment]
Theano code seems to be working, but I can't debug the tensorflow code

**Comment by EderSantana** [comment]
@yaringal there is a `None`, defined in the test and TF deconv does not accept that. I think I know where it is... let me try to fix this. I'll make a PR to your PR

**Comment by EderSantana** [comment]
@yaringal I made you a PR,I think it should fix it

**Comment by EderSantana** [comment]
@yaringal sorry for being distracted, I "fixed" the wrong test last time. I made another corrections.

I can run the tests locally right now. If this doesn't work, I'll get back to this tomorrow

**Comment by EderSantana** [comment]
yay!!! we are good here @fchollet 
let us know what do you think, specially regarding the `output_shape` inferecen

File: keras/layers/convolutional.py [hunk 6]
```diff
@@ -1261,5 +1351,7 @@ def get_config(self):
 Conv1D = Convolution1D
 Conv2D = Convolution2D
 Conv3D = Convolution3D
+Deconv2D = Deconvolution2D
+TransposedConvolution2D = Deconvolution2D
```
Comment (fchollet) [inline keras/layers/convolutional.py]: I don't think we need this alias, it would increase confusion.

File: keras/utils/test_utils.py [hunk 7]
```diff
@@ -36,7 +36,7 @@ def get_test_data(nb_train=1000, nb_test=500, input_shape=(10,),
 
 
 def layer_test(layer_cls, kwargs={}, input_shape=None, input_dtype=None,
-               input_data=None, expected_output=None, expected_output_dtype=None):
+               input_data=None, expected_output=None, expected_output_dtype=None, with_None=True):
```
Comment (fchollet) [inline keras/utils/test_utils.py]: `with_None` is non-standard style, and not descriptive of what it's supposed to do (also undocumented).

Comment (EderSantana) [inline keras/utils/test_utils.py]: yeah, I was thinking about that... but the problem is deconv in TF doesn't handle `None` in the shape dim. the test has to specify all dimensions. so this `None` refers to the number of samples
how should we call it instead?

Comment (fchollet) [inline keras/utils/test_utils.py]: `fixed_batch_size=False`

Comment (EderSantana) [inline keras/utils/test_utils.py]: perfect!

**Comment by EderSantana** [comment]
the failure looks weird, doesn't seem related to the PR, does it?

**Comment by yaringal** [comment]
Build still seems to fail with error 
`The command "if [[ "$TRAVIS_PYTHON_VERSION" == "2.7" ]]; then wget https://repo.continuum.io/miniconda/Miniconda-latest-Linux-x86_64.sh -O miniconda.sh; else wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh; fi" failed and exited with 8 during .`
`Your build has been stopped.`

**Comment by lukovkin** [comment]
@EderSantana @yaringal I've tried to do an output_shape inference for this PR.
Any advice how can I make a PR with those changes into this PR? ;)

**Comment by yaringal** [comment]
@lukovkin I think you can do a pull request at https://github.com/yaringal/keras/commits/master which I can merge into this PR

**Comment by fchollet** [comment]
It seems the tests are failing due to an issue with Travis. I would expect it to go away; if it doesn't me may have to update our Travis config.

**Comment by fchollet** [comment]
(Travis is fixed and the tests are passing)

**Comment by EderSantana** [comment]
@fchollet do you have any further suggestions for changes?

File: keras/backend/tensorflow_backend.py [hunk 8]
```diff
@@ -1168,6 +1168,12 @@ def l2_normalize(x, axis):
 
 # CONVOLUTIONS
 
+def _preprocess_deconv_output_shape(sh, dim_ordering):
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: I think using `shape` would be more in line with the general coding style

File: keras/backend/tensorflow_backend.py [hunk 9]
```diff
@@ -1168,6 +1168,12 @@ def l2_normalize(x, axis):
...
+    if dim_ordering == "th":
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Use `'` not `"`, as in the rest of the file

File: keras/backend/tensorflow_backend.py [hunk 10]
```diff
@@ -1293,11 +1299,12 @@ def deconv2d(x, kernel, output_shape, strides=(1, 1),
         raise Exception('Unknown dim_ordering ' + str(dim_ordering))
 
     x = _preprocess_conv2d_input(x, dim_ordering)
+    output_shape = _preprocess_deconv_output_shape(output_shape, dim_ordering)
     kernel = _preprocess_conv2d_kernel(kernel, dim_ordering)
+    kernel = tf.transpose(kernel, (0, 1, 3, 2))  # tranpose kernel chanels
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Comment not necessary (also, has a typo)

File: keras/backend/theano_backend.py [hunk 11]
```diff
@@ -912,58 +912,106 @@ def l2_normalize(x, axis):
...
+    if dim_ordering == 'tf':
+        if filter_shape:
+            filter_shape = (filter_shape[3], filter_shape[2],
+                            filter_shape[0], filter_shape[1])
     if filter_shape is not None:
         filter_shape = tuple(int_or_none(v) for v in filter_shape)
+    return filter_shape
+
+
+def _postprocess_border_mode(conv_out, x, border_mode, np_kernel, strides):
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: `_postprocess_border_mode` should be part of  `_postprocess_conv2d_output`, it doesn't need to be its own function.

File: keras/backend/tensorflow_backend.py [hunk 12]
```diff
@@ -1293,11 +1299,12 @@ def deconv2d(x, kernel, output_shape, strides=(1, 1),
...
     padding = _preprocess_border_mode(border_mode)
     strides = (1,) + strides + (1,)
 
-    # TODO: pre-process output_shape if dim_ordering == th
     x = tf.nn.conv2d_transpose(x, kernel, output_shape, strides,
                                padding=padding)
     return _postprocess_conv2d_output(x, dim_ordering)
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: At this point it would be fine to do the dimshuffle without the call the `_postprocess_conv2d_output`

File: keras/backend/theano_backend.py [hunk 13]
```diff
@@ -912,58 +912,106 @@ def l2_normalize(x, axis):
...
-        if image_shape:
-            image_shape = (image_shape[0], image_shape[3],
-                           image_shape[1], image_shape[2])
-        if filter_shape:
-            filter_shape = (filter_shape[3], filter_shape[2],
-                            filter_shape[0], filter_shape[1])
+    return kernel
 
+
+def _preprocess_border_mode(border_mode, kernel):
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Processing of  `np_kernel` should be removed from this function

File: keras/backend/theano_backend.py [hunk 14]
```diff
@@ -980,22 +1028,46 @@ def int_or_none(value):
...
+        in inputs/kernels/ouputs.
+    '''
+    flip_filters = False
+    if dim_ordering not in {'th', 'tf'}:
+        raise Exception('Unknown dim_ordering ' + str(dim_ordering))
+
+    x = _preprocess_conv2d_input(x, dim_ordering)
+    kernel = _preprocess_conv2d_kernel(kernel, dim_ordering)
+    kernel = kernel.dimshuffle((1, 0, 2, 3))
+    th_border_mode, np_kernel = _preprocess_border_mode(border_mode, kernel)
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: At this point it would be fine to have `np_kernel = kernel.eval()`

File: keras/layers/convolutional.py [hunk 15]
```diff
@@ -379,6 +379,96 @@ def get_config(self):
...
+        if self.dim_ordering == 'th':
+            return (input_shape[0], self.nb_filter, rows, cols)
+        elif self.dim_ordering == 'tf':
+            return (input_shape[0], rows, cols, self.nb_filter)
+        else:
+            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
+        # return self.output_shape_
+
+    def call(self, x, mask=None):
+        # output_shape = self.get_output_shape_for(x.shape)
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Comment does not seem necessary (leftover?)

File: keras/layers/convolutional.py [hunk 16]
```diff
@@ -379,6 +379,96 @@ def get_config(self):
...
+                  'activation': self.activation.__name__,
+                  'border_mode': self.border_mode,
+                  'subsample': self.subsample,
+                  'dim_ordering': self.dim_ordering,
+                  'W_regularizer': self.W_regularizer.get_config() if self.W_regularizer else None,
+                  'b_regularizer': self.b_regularizer.get_config() if self.b_regularizer else None,
+                  'activity_regularizer': self.activity_regularizer.get_config() if self.activity_regularizer else None,
+                  'W_constraint': self.W_constraint.get_config() if self.W_constraint else None,
+                  'b_constraint': self.b_constraint.get_config() if self.b_constraint else None,
+                  'bias': self.bias}
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Most of this stuff is included in the config of  `super` and thus does not need to figure here.

**Comment by fchollet** [comment]
Yes, I added a series of comments on the code.

**Comment by lukovkin** [comment]
@yaringal I've created a PR in your fork.

**Comment by yaringal** [comment]
@lukovkin thanks - let me know when it's working with theano and I'll merge (I tried to do something similar a few days ago but theano complained that conv cannot take symbolic variables)

**Comment by yaringal** [comment]
@EderSantana are you working on the changes @fchollet suggested or shall I?
